pandas/tests/test_frame.py

```diff
@@ -12385,6 +12404,93 @@ def verify(df):
                 for col in ['4th', '5th']:
                     verify(udf[col])
 
+        # GH7403
+        df = pd.DataFrame({'A': list('aaaabbbb'),'B':range(8), 'C':range(8)})
+        df.iloc[3, 1] = np.NaN
+        left = df.set_index(['A', 'B']).unstack(0)
+
+        vals = [[3, 0, 1, 2, nan, nan, nan, nan],
+                [nan, nan, nan, nan, 4, 5, 6, 7]]
+        vals = list(map(list, zip(*vals)))
+        idx = Index([nan, 0, 1, 2, 4, 5, 6, 7], name='B')
```

Comment:
@behzadnouri, just to clarify, I don't doubt that within the unstacking code the conversion from ints to floats happens in ``Index.__init__()``, but that doesn't change the fact that in such cases ``unstack()`` has the effect of converting ints to floats.

The reason I think this is a potential problem (which predates you, I'm sure, and is also present in ``stack()``), is that if someone then wants to compare the values in the ``Index`` of the stacked/unstacked object to the values in the corresponding level of ``MultiIndex`` of the original ``DataFrame``, they won't match.